src/ipc/mmap/synchronizer.rs

```diff
@@ -0,0 +1,69 @@
+use crate::ipc::{ChannelTrait, ChannelWithSerialize};
+use mmap_sync::{instance::InstanceVersion, synchronizer::Synchronizer};
+use std::{error::Error, time::Duration};
+
+pub struct MmapSynchronizer {
+    reader: Synchronizer,
+    writer: Synchronizer,
+    reader_state: InstanceVersion,
+}
+
+impl MmapSynchronizer {
+    pub fn conn() -> Self {
+        let mut connector = Self::new("./pg_vectors/_mmap_listen", false);
+        connector.write(&vec![0u8; 1]).expect("Failed to write.");
+        let num: usize = (&mut connector as &mut dyn ChannelTrait)
+            .recv()
+            .expect("Failed to recv.");
+        let path = format!("./pg_vectors/_mmap_{num}");
+        Self::new(&path, false)
+    }
+
+    pub fn new(path: &str, is_server: bool) -> Self {
+        // The filename represents which side reads it.
+        let mut reader_path = path.to_string();
+        let mut writer_path = path.to_string();
+        if is_server {
+            reader_path.push_str("_server");
+            writer_path.push_str("_client");
+        } else {
+            writer_path.push_str("_client");
+            reader_path.push_str("_server");
+        }
+        let mut reader = Synchronizer::new(reader_path.as_ref());
+        let writer = Synchronizer::new(writer_path.as_ref());
+
+        // The reader need to write something first to initialize the mmap file and state.
+        reader
+            .write(&[0u8; 1], Duration::from_secs(1))
+            .expect("Failed to write.");
+        unsafe { reader.read::<[u8; 1]>(false) }.expect("Failed to read.");
```

Comment:
I think the writing is done if no exception raised? Because the mmap-sync is sync

Reply:
> I think the writing is done if no exception raised?

Yes. But the other side don't know when the writing is done.

Reply:
I think the reader need to spawn a spin-loop thread to actively read from mmap to a buffer, and the ChannelTrait read is designed to read from the buffer when needed.

Reply:
> Is this the common pattern for stream based interface? For socket interface, receiver also don't know when the write operation is done by the writer?

For socket interface, the reader will be blocked until the write operation is done by the writer. But `mmap-sync` won't block anything. For read operations other than the first read, we can use spin-loop to block it mannually. But for the first time, it will directly panic.

> I think the reader need to spawn a spin-loop thread to actively read from mmap to a buffer

Using the way, we will give up the advantage of zero-copy with mmap.

Reply:
> For read operations other than the first read, we can use spin-loop to block it mannually.

Since the mmap is the only IPC data link, I dont think there are any method for writer to "notify" the receiver that the writting is done.

> Using the way, we will give up the advantage of zero-copy with mmap.

The writer doesn't check if the sent data is consumed, so the next data will overwrite the previous one and causes "lossing packet"

Reply:
> Since the mmap is the only IPC data link, I dont think there are any method for writer to "notify" the receiver that the writting is done.

We can store extra state data in an additional mmap file. Because the memory is shared between 2 processes, all sync primitives can be used, like locks, etc.

> The writer doesn't check if the sent data is consumed, so the next data will overwrite the previous one and causes "lossing packet".

We can use write data in different place, also it's possible to check if the sent data is consumed. `mmap-sync` use [`RCU`](https://www.kernel.org/doc/html/next/RCU/whatisRCU.html) to implement similar function.

Reply:
Thanks for your explanation. @usamoi will implement this feature, and we're single producer, single consumer so it can be much easier. 
Spin-lock is not a good solution since it consumes lots of cpu resources.

Reply:
Seems https://github.com/servo/ipc-channel can be a good [baseline](https://github.com/dunnock/ipc-bench/tree/master)